code/workspaces/infrastructure-api/src/dataaccess/stacksRepository.js

```diff
@@ -6,34 +6,54 @@ function Stack() {
 
 function getAll(user) {
   return Stack()
-    .find().filterByUser(user).exec();
+    .find()
+    .filterByUser(user)
+    .exec();
+}
+
+function getAllByProject(projectKey, user) {
+  return Stack()
+    .find()
+    .filterByProject(projectKey)
+    .filterByUser(user)
+    .exec();
 }
 
-function getAllByCategory(user, category) {
+function getAllByCategory(projectKey, user, category) {
   return Stack()
-    .find({ category }).filterByUser(user).exec();
+    .find({ category })
+    .filterByProject(projectKey)
+    .filterByUser(user)
+    .exec();
 }
 
-function getAllByVolumeMount(user, mount) {
+function getAllByVolumeMount(projectKey, user, mount) {
   // Searches All Users
   return Stack()
-    .find({ volumeMount: mount }).exec();
+    .find({ volumeMount: mount })
+    .filterByProject(projectKey)
+    .exec();
 }
 
 function getOneById(user, id) {
   return Stack()
-    .findOne({ _id: id }).filterOneByUser(user).exec();
+    .findOne({ _id: id })
+    .filterOneByUser(user)
+    .exec();
 }
 
-function getOneByName(user, name) {
+function getOneByName(projectKey, user, name) {
   // Searches All Users
   return Stack()
-    .findOne({ name }).exec();
+    .findOne({ name })
+    .filterOneByProject(projectKey)
+    .exec();
 }
 
-function createOrUpdate(user, stack) {
+function createOrUpdate(projectKey, user, stack) {
   return Stack()
     .find()
+    .filterByProject(projectKey)
```

Comment:
Is this function really just update?

Reply:
It is an upsert request so we still need to check for a previous stack to update prior to inserting a new one. I have successfully inserted and updated new stacks during my testing.

Reply:
Thanks, I missed that bit.